packages/core/src/CSL/cslToCore.ts

```diff
@@ -81,17 +82,158 @@ export const txInputs = (inputs: CSL.TransactionInputs, address?: Cardano.Addres
+        port: relayByAddress.port()
+      });
+    }
+    if (relayByName) {
+      // RelayByName
+      result.push({
+        __typename: 'RelayByName',
+        hostname: relayByName.dns_name().record(),
+        port: relayByName.port()
+      });
+    }
+    if (relayByNameMultihost) {
+      // RelayByNameMultihost
+      result.push({
+        __typename: 'RelayByNameMultihost',
+        dnsName: relayByNameMultihost.dns_name().record()
+      });
+    }
+  }
+  return result;
+};
+
+const createCardanoOwners = (owners: CSL.Ed25519KeyHashes): Cardano.RewardAccount[] => {
+  const result: Cardano.RewardAccount[] = [];
+  for (let i = 0; i < owners.len(); i++) {
+    const owner = owners.get(i);
+    result.push(Cardano.RewardAccount(Buffer.from(owner.to_bytes()).toString()));
+  }
+  return result;
+};
+
+const poolRegistration = (certificate: CSL.PoolRegistration): Cardano.PoolRegistrationCertificate => {
+  const { reward_account, pledge, cost, margin, pool_metadata, pool_owners, relays, vrf_keyhash } =
+    certificate.pool_params();
+  return {
+    __typename: Cardano.CertificateType.PoolRegistration,
+    poolParameters: {
+      cost: BigInt(cost.toString()),
+      id: '???', // Operator?
+      margin: {
+        denominator: Number(margin().denominator().to_str()),
+        numerator: Number(margin().numerator().to_str())
+      },
+
+      pledge: BigInt(pledge.toString()),
+      rewardAccount: Cardano.RewardAccount(reward_account().to_address().to_bech32()),
+      ...(pool_metadata() && {
+        metadataJson: {
+          hash: pool_metadata()?.pool_metadata_hash(),
+          url: pool_metadata()?.url()
+        }
+      }),
+      owners: createCardanoOwners(pool_owners()),
+      relays: createCardanoRelays(relays()),
+      vrf: Cardano.VrfVkHex(Buffer.from(vrf_keyhash().to_bytes()).toString())
+    }
+  };
+};
+
+const poolRetirement = (certificate: CSL.PoolRetirement): Cardano.PoolRetirementCertificate => ({
+  __typename: Cardano.CertificateType.PoolRetirement,
+  epoch: certificate.epoch(),
+  poolId: Cardano.PoolId(certificate.pool_keyhash().toString())
+});
+
+const genesisKeyDelegaation = (certificate: CSL.GenesisKeyDelegation): Cardano.GenesisKeyDelegationCertificate => ({
+  __typename: Cardano.CertificateType.GenesisKeyDelegation,
+  genesisDelegateHash: Cardano.util.Hash32ByteBase16(
+    Buffer.from(certificate.genesis_delegate_hash().to_bytes()).toString()
+  ),
+  genesisHash: Cardano.util.Hash32ByteBase16(Buffer.from(certificate.genesishash().to_bytes()).toString()),
+  vrfKeyHash: Cardano.util.Hash32ByteBase16(Buffer.from(certificate.vrf_keyhash().to_bytes()).toString())
+});
+
+const mir = (certificate: CSL.MoveInstantaneousRewardsCert): Cardano.MirCertificate => {
+  const potNumber = certificate.move_instantaneous_reward().pot();
+
+  return {
+    __typename: Cardano.CertificateType.MIR,
+    pot: Object.values(Cardano.MirCertificatePot)[potNumber],
+    quantity: '???',
```

Comment:
Seems this certificate also has `kind()`, so I'm guessing one of those is 'quantity' and the other one is 'rewardAccount'. That would make our type incorrect, as we require both, while CSL suggests they're mutually exclusive. @james-iohk could you please confirm this?

Reply:
Just noticed that `coreToCsl` doesn't support this certificate too. Let's just scope this out to a separate task to implement both conversions.

Reply:
removed, todo added